torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -72,6 +73,74 @@ std::pair<xla::XlaOp, xla::XlaOp> DotBroadcast(const xla::XlaOp& lhs,
 
 }  // namespace
 
+std::vector<xla::XlaOp> CreateKthValue(const xla::XlaOp& input, xla::int64 k,
+                                       xla::int64 dim, bool keepdim) {
+  // Here 'k' is 1 based (1...).
+  xla::Shape shape = XlaHelpers::ShapeOfXlaOp(input);
+  XLA_CHECK_LE(k, shape.dimensions(dim));
+  xla::Shape iota_shape =
+      xla::ShapeUtil::MakeShape(xla::PrimitiveType::S32, shape.dimensions());
+  xla::XlaOp iota = xla::Iota(input.builder(), iota_shape, dim);
+  xla::XlaOp sort_result = xla::Sort(
+      {input, iota},
+      xla::CreateScalarLtComputation(
+          {shape.element_type(), xla::PrimitiveType::S32}, input.builder()),
+      dim);
+
+  std::vector<xla::int64> start_indices(shape.rank(), 0);
+  start_indices[dim] = k - 1;
+  std::vector<xla::int64> limit_indices(shape.dimensions().begin(),
+                                        shape.dimensions().end());
+  limit_indices[dim] = k;
+  std::vector<xla::int64> strides(shape.rank(), 1);
+
+  xla::XlaOp values = xla::Slice(xla::GetTupleElement(sort_result, 0),
+                                 start_indices, limit_indices, strides);
+  xla::XlaOp indices = xla::Slice(xla::GetTupleElement(sort_result, 1),
+                                  start_indices, limit_indices, strides);
+  if (!keepdim) {
+    auto reshape_sizes = XlaHelpers::DropDimensions(shape.dimensions(), {dim});
+    values = xla::Reshape(values, reshape_sizes);
+    indices = xla::Reshape(indices, reshape_sizes);
+  }
+  // aten::kthvalue() wants Long tensors as indices.
+  return {values, xla::ConvertElementType(indices, xla::PrimitiveType::S64)};
+}
+
+std::vector<xla::XlaOp> CreateTopK(const xla::XlaOp& input, xla::int64 k,
+                                   xla::int64 dim, bool largest,
+                                   bool /* sorted */) {
```

Comment:
What is `sorted` supposed to do?

Reply:
I assumed it was an optimization to return in order, or any order.
But now that I think about it, could be in-order (the default), or in the native order (sorted=false).
Native order sucks because we have to either throw away the sort result and do a gather on the sorted indices, or shuffle the sort result back into native order.
Let me pun to CPU in case of not-sorted until I work my way on the best approach.

Reply:
Yikes, yes, let's not handle not-sorted until we figure it out.

Reply:
Actually, seems a bit "whatever".

```
>>> x = torch.randn(10)
>>> print(x)
tensor([-1.3195, -0.3020,  0.2261, -2.0767, -1.5631,  0.3084, -1.4165, -0.2338,
         0.8787,  0.5441])
>>> y = torch.topk(x, 2, 0, largest=True, sorted=True)
>>> print(y[0])
tensor([0.8787, 0.5441])
>>> y = torch.topk(x, 2, 0, largest=True, sorted=False)
>>> print(y[0])
tensor([0.5441, 0.8787])
```

With sorted=False I would have expected to see [0.8787, 0.5441] (the native order).
I think sorted=False it is really whatever, and we might be able to remove the pun-to-CPU.